pkg/kocache/offlinectrl.go

```diff
@@ -0,0 +1,56 @@
+package kocache
+
+import (
+	"context"
+
+	"github.com/pkg/errors"
+	"github.com/stackrox/rox/pkg/sync"
+)
+
+var errCentralUnreachable = errors.New("central is currently unreachable")
+
+// offlineCtrl exists to ensure consistency between variables being related to offline state
+type offlineCtrl struct {
+	parentCtx context.Context
+	// onlineCtx will be canceled when connectivity to central is lost
+	onlineCtx       context.Context
+	onlineCtxCancel context.CancelCauseFunc
+	ctxMutex        *sync.Mutex
+}
+
+func newOfflineCtrl(parentCtx context.Context, startOnline bool) *offlineCtrl {
+	if parentCtx == nil {
+		parentCtx = context.Background()
+	}
+	onlineCtx, onlineCtxCancel := context.WithCancelCause(parentCtx)
+	oc := &offlineCtrl{
+		parentCtx:       parentCtx,
+		onlineCtx:       onlineCtx,
+		onlineCtxCancel: onlineCtxCancel,
+		ctxMutex:        &sync.Mutex{},
+	}
+	if !startOnline {
+		oc.GoOffline()
+	}
+	return oc
+}
+
+func (o *offlineCtrl) IsOnline() bool {
+	return o.onlineCtx.Err() == nil
```

Comment:
Could you elaborate a bit more? 
I understand that there may be a situation where the error is not nil and we should still stay online. What would that mean?

Reply:
I will merge the current state. If we decide to improve the part from this comment, then we can do this in a followup.

Reply:
You are assuming that contexts will always be cancelled with an error. That's because we create our own context in this component to control whether the connection is online or offline. 

What you want to check here is if Sensor is currently connected or not. Since the statement above is true in the current state of the code, whenever a context has an error, it's safe to assume that the connection is broken.

The problem is that this only works because the context being created here is inheriting from `context.Background()` and therefore it won't never be cancelled. If for whatever reason we start inheriting from a different context that _could_ be cancelled without an error, this check will return `true` even when the context was cancelled. 

This would mean that functions that are asking `IsOnline` will get `true` even if the context being passed along is already cancelled. 

I wonder why aren't you checking `o.onlineContext.Done()`? Wouldn't it be more stable?

Reply:
This is not a problem then. Thanks for clarifying. :-)